cpp/src/distance/distance.cu

```diff
@@ -914,21 +1232,29 @@ instantiate_raft_distance_distance(cuvs::distance::DistanceType::RusselRaoExpand
                                    double,
                                    raft::layout_f_contiguous,
                                    int);
+instantiate_raft_distance_distance(cuvs::distance::DistanceType::RusselRaoExpanded,
+                                   half,
+                                   float,
+                                   float,
+                                   raft::layout_f_contiguous,
+                                   int);
 
 #undef instantiate_raft_distance_distance
 
-#define instantiate_raft_distance_pairwise_distance(DataT, layout, IdxT) \
-  template void cuvs::distance::pairwise_distance(                       \
-    raft::resources const& handle,                                       \
-    raft::device_matrix_view<const DataT, IdxT, layout> const x,         \
-    raft::device_matrix_view<const DataT, IdxT, layout> const y,         \
-    raft::device_matrix_view<DataT, IdxT, layout> dist,                  \
-    cuvs::distance::DistanceType metric,                                 \
-    DataT metric_arg)
+#define instantiate_raft_distance_pairwise_distance(DataT, layout, IdxT, OutT) \
+  template void cuvs::distance::pairwise_distance(                             \
+    raft::resources const& handle,                                             \
+    raft::device_matrix_view<const DataT, IdxT, layout> const x,               \
+    raft::device_matrix_view<const DataT, IdxT, layout> const y,               \
+    raft::device_matrix_view<OutT, IdxT, layout> dist,                         \
```

Comment:
Why do we need a new template param for this? Is this absolutely necessary or is this just a convenience? 

I understand you were able to drop the binary size down in this PR by fixing the pairwise distance instnatiations, but that fix should be decoupled from your additions here. In other words- if we could drop the binary size by 100mb or more just by fixing exiting instantiations, I'd rather not add new intantiations in unecessarily that add to the binary size. Can these be at all avoided using the existing template types? We have a lot of them already...

Reply:
Yeah, I'm afraid it's necessary if we want to support a distance type different from the input type(like half -> float) unless we make large-scale refactoring on this part of code because all of the APIs assume inputs and output share the same type before.